lighthouse-core/gather/gather-runner.js

```diff
@@ -196,18 +196,30 @@ class GatherRunner {
 
       // Run each pass
       .then(_ => {
-        return passes.reduce((chain, config) => {
-          const runOptions = Object.assign({}, options, {config});
+        // If the main document redirects, we'll update this to keep track
+        let urlAfterRedirects;
+        return passes.reduce((chain, config, passIndex) => {
+          let runOptions;
           return chain
+              .then(_ => {
+                runOptions = Object.assign({}, options, {config});
+              })
               .then(_ => GatherRunner.beforePass(runOptions))
               .then(_ => GatherRunner.pass(runOptions))
               .then(_ => GatherRunner.afterPass(runOptions))
               .then(loadData => {
                 // Merge pass trace and network data into tracingData.
                 config.trace && Object.assign(tracingData.traces, loadData.traces);
                 config.network && (tracingData.networkRecords = loadData.networkRecords);
+              })
+              .then(_ => {
+                if (passIndex === 0) {
+                  urlAfterRedirects = runOptions.url;
+                }
               });
-        }, Promise.resolve());
+        }, Promise.resolve()).then(_ => {
+          options.url = urlAfterRedirects;
+        });
```

Comment:
since you no longer alter the options object in this loop you can simplify this again:
```js
.then(_ => {
  // If the main document redirects, we'll update this to keep track
  let urlAfterRedirects;
  return passes.reduce((chain, config, passIndex) => {
    const runOptions = Object.assign({}, options, {config});
    return chain
      .then(_ => GatherRunner.beforePass(runOptions))
      .then(_ => GatherRunner.pass(runOptions))
      .then(_ => GatherRunner.afterPass(runOptions))
      .then(loadData => {
        // Merge pass trace and network data into tracingData.
        config.trace && Object.assign(tracingData.traces, loadData.traces);
        config.network && (tracingData.networkRecords = loadData.networkRecords);
        if (passIndex === 0) {
          urlAfterRedirects = runOptions.url;
        }
      });
  }, Promise.resolve()).then(_ => {
    options.url = urlAfterRedirects;
  });
})
```
(technically you could even set `options.url` to `runOptions.url` in that `passIndex` conditional since each pass already has its own copy of options but that might be confusing)